QuickFIXn/SocketReader.cs

```diff
@@ -71,46 +75,44 @@ public void Read()
         /// <exception cref="System.Net.Sockets.SocketException">On connection reset</exception>
         protected virtual int ReadSome(byte[] buffer, int timeoutMilliseconds)
         {
-            // NOTE: THIS FUNCTION IS EXACTLY THE SAME AS THE ONE IN SocketInitiatorThread.
-            // Any changes made here should also be made there.
-            try
-            {
+            try {
                 // Begin read if it is not already started
-                _currentReadRequest ??= _stream.BeginRead(buffer, 0, buffer.Length, null, null);
+                _currentReadTask ??= _stream.ReadAsync(buffer, 0, buffer.Length, _readCancellationTokenSource.Token);
 
                 // Wait for it to complete (given timeout)
-                _currentReadRequest.AsyncWaitHandle.WaitOne(timeoutMilliseconds);
+                _currentReadTask.Wait(timeoutMilliseconds);
 
-                if (_currentReadRequest.IsCompleted)
-                {
-                    // Make sure to set _currentReadRequest to before retreiving result
-                    // so a new read can be started next time even if an exception is thrown
-                    var request = _currentReadRequest;
-                    _currentReadRequest = null;
+                if (_currentReadTask.IsCompleted) {
+                    // Dispose/nullify currentReadTask *before* retreiving .Result.
+                    //   Accessting .Result can throw an exception, so we need to reset currentReadTask
+                    //   first, to set us up for the next read even if an exception is thrown.
+                    Task<int>? request = _currentReadTask;
+                    _currentReadTask.Dispose();
```

Comment:
Disposing Tasks is [unnecessary](https://devblogs.microsoft.com/pfxteam/do-i-need-to-dispose-of-tasks/). And despite the blog confirming that calling `Result` afterwards should be OK, it still feels a bit strange to call `Result` after `Dispose`. I would suggest removing this line.

Reply:
I'm realizing now that I could just call `request.Dispose()` after `request.Result`.

But of course, if I don't need to Dispose at all, then I won't.